src/backing/status/status_or.h

```diff
@@ -0,0 +1,291 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Modifications copyright 2020 Google LLC:
+ *
+ *    - Renamed constants to match k-naming
+ *    - Replaced references to StringPiece datatype with std::string
+ *    - Changed Status.message and Status.error_message usage to reflect change
+ *      from StringPiece datatype to std::string
+ *
+ * Modifications licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// StatusOr<T> is the union of a Status object and a T
+// object. StatusOr models the concept of an object that is either a
+// usable value, or an error Status explaining why such a value is
+// not present. To this end, StatusOr<T> does not allow its Status
+// value to be Status::OK. Further, StatusOr<T*> does not allow the
+// contained pointer to be nullptr.
+//
+// The primary use-case for StatusOr<T> is as the return value of a
+// function which may fail.
+//
+// Example client usage for a StatusOr<T>, where T is not a pointer:
+//
+//  StatusOr<float> result = DoBigCalculationThatCouldFail();
+//  if (result.ok()) {
+//    float answer = result.ValueOrDie();
+//    printf("Big calculation yielded: %f", answer);
+//  } else {
+//    LOG(ERROR) << result.status();
+//  }
+//
+// Example client usage for a StatusOr<T*>:
+//
+//  StatusOr<Foo*> result = FooFactory::MakeNewFoo(arg);
+//  if (result.ok()) {
+//    std::unique_ptr<Foo> foo(result.ValueOrDie());
+//    foo->DoSomethingCool();
+//  } else {
+//    LOG(ERROR) << result.status();
+//  }
+//
+// Example client usage for a StatusOr<std::unique_ptr<T>>:
+//
+//  StatusOr<std::unique_ptr<Foo>> result = FooFactory::MakeNewFoo(arg);
+//  if (result.ok()) {
+//    std::unique_ptr<Foo> foo = result.ValueOrDie();
+//    foo->DoSomethingCool();
+//  } else {
+//    LOG(ERROR) << result.status();
+//  }
+//
+// Example factory implementation returning StatusOr<T*>:
+//
+//  StatusOr<Foo*> FooFactory::MakeNewFoo(int arg) {
+//    if (arg <= 0) {
+//      return ::Status(::StatusCode::kInvalidArgument,
+//                            "Arg must be positive");
+//    } else {
+//      return new Foo(arg);
+//    }
+//  }
+//
+
+#ifndef KMSENGINE_BACKING_STATUS_STATUS_OR_H_
+#define KMSENGINE_BACKING_STATUS_STATUS_OR_H_
+
+#include "src/backing/status/status_or.h"
+
+#include <new>
+#include <string>
+#include <utility>
+
+#include "src/backing/status/status.h"
+#include "src/backing/status/status_code.h"
+
+namespace kmsengine {
+namespace backing {
```

Comment:
Where does the name "backing" come from?

I wouldn't be opposed to just leaving the statuses in the common namespace.

Reply:
It's the name for the `libkms` part of the codebase. Given the previous discussion on not including "lib" in the namespace names, I figured `namespace kmsengine { namespace kms { ... } }` might be confusing and tried to pick something else.

I think leaving in the common namespace is a good idea (as I've found while trying to use these types) - I've just been trying to make a clear distinction between which parts of the codebase are only allowed to depend on OpenSSL and which parts of the codebase are allowed to depend on other dependencies (that aren't OpenSSL).